Generating *_Generated_RestServiceProxy_.java fails when List<T> is expected from REST service

Hi!

I have the following REST service on my GWT client:
```java
package test.client.gwt.client.service.rest;

import java.util.List;

import javax.ws.rs.Path;
import javax.ws.rs.QueryParam;

import org.fusesource.restygwt.client.JSONP;
import org.fusesource.restygwt.client.MethodCallback;
import org.fusesource.restygwt.client.RestService;

import test.common.dto.User;

public interface UserService extends RestService {

	@Path("http://localhost:8080/test.server.rest/users")
	@JSONP
	void getUserList(
			@QueryParam("offset") Integer offset,
			@QueryParam("limit") Integer limit,
			MethodCallback<List<User>> callback);
}
```

where `User` is defined as

```java
package test.common.dto;

import java.math.BigInteger;

public class User {

	private BigInteger id;

	private String firstName;

	private String secondName;

	public User() {}

	public User(String firstName, String secondName) {
		this.firstName = firstName;
		this.secondName = secondName;
	}

	public String getFirstName() {
		return firstName;
	}

	public void setFirstName(String firstName) {
		this.firstName = firstName;
	}

	public String getSecondName() {
		return secondName;
	}

	public void setSecondName(String secondName) {
		this.secondName = secondName;
	}

	public BigInteger getId() {
		return id;
	}

	public void setId(BigInteger id) {
		this.id = id;
	}

}
```

When trying to compile my GWT application I get the folliwing error:

```
[INFO] --- gwt-maven-plugin:2.7.0:compile (default) @ test.client.gwt ---
[INFO] Compiling module test.client.gwt.testClientGWT
[INFO]    Adding '2' new generated units
[INFO]       See snapshot: C:\Windows\TEMP\test.client.gwt.client.service.rest.UserService_Generated_RestServiceProxy_7823676910508270362.java
[INFO]       Ignored 1 unit with compilation errors in first pass.
[INFO] Compile with -strict or with -logLevel set to TRACE or DEBUG to see all errors.
[INFO]    Tracing compile failure path for type 'test.client.gwt.client.service.rest.UserService_Generated_RestServiceProxy_'
[INFO]       [ERROR] Errors in 'C:\Work\test\workspace_4\test.client.gwt\target\.generated\test\client\gwt\client\service\rest\UserService_Generated_RestServiceProxy_.java'
[INFO]          [ERROR] Line 38: The method getJavaScriptObject() is undefined for the type JSONValue
[INFO]    [ERROR] Hint: Check that the type name 'test.client.gwt.client.service.rest.UserService_Generated_RestServiceProxy_' is really what you meant
[INFO]    [ERROR] Hint: Check that your classpath includes all required source roots
[INFO]    [ERROR] Errors in 'test/client/gwt/client/application/userlist/UserListPresenter.java'
[INFO]       [ERROR] Line 27: Rebind result 'test.client.gwt.client.service.rest.UserService_Generated_RestServiceProxy_' could not be found
```


The generated class that have errors looks in this way:
```java
package test.client.gwt.client.service.rest;

public class UserService_Generated_RestServiceProxy_ implements test.client.gwt.client.service.rest.UserService,
		org.fusesource.restygwt.client.RestServiceProxy {
	private org.fusesource.restygwt.client.Resource resource = null;

	@Override
	public void setResource(org.fusesource.restygwt.client.Resource resource) {
		this.resource = resource;
	}

	@Override
	public org.fusesource.restygwt.client.Resource getResource() {
		if (this.resource == null) {
			String serviceRoot = org.fusesource.restygwt.client.Defaults.getServiceRoot();
			this.resource = new org.fusesource.restygwt.client.Resource(serviceRoot);
		}
		return this.resource;
	}

	private org.fusesource.restygwt.client.Dispatcher dispatcher = null;

	@Override
	public void setDispatcher(org.fusesource.restygwt.client.Dispatcher dispatcher) {
		this.dispatcher = dispatcher;
	}

	@Override
	public org.fusesource.restygwt.client.Dispatcher getDispatcher() {
		return this.dispatcher;
	}

	@Override
	public void getUserList(java.lang.Integer offset, java.lang.Integer limit,
			org.fusesource.restygwt.client.MethodCallback<java.util.List<test.common.dto.User>> callback) {
		final java.lang.Integer final_offset = offset;
		final java.lang.Integer final_limit = limit;
		final org.fusesource.restygwt.client.Method __method =
				getResource()
						.resolve("http://localhost:8080/test.server.rest/users")
						.addQueryParam("offset", (offset != null ? offset.toString() : null))
						.addQueryParam("limit", (limit != null ? limit.toString() : null))
						.jsonp();
		__method.setDispatcher(this.dispatcher);
		((org.fusesource.restygwt.client.JsonpMethod) __method)
				.send(new org.fusesource.restygwt.client.AbstractAsyncCallback<java.util.List<test.common.dto.User>>(
						(org.fusesource.restygwt.client.JsonpMethod) __method, callback) {
					@Override
					protected java.util.List<test.common.dto.User> parseResult(
							com.google.gwt.json.client.JSONValue result) throws Exception {
						try {
							result = new com.google.gwt.json.client.JSONArray(result.getJavaScriptObject());
							return org.fusesource.restygwt.client.AbstractJsonEncoderDecoder.toList(result,
									test.common.dto.User_Generated_JsonEncoderDecoder_.INSTANCE);
						} catch (Throwable __e) {
							throw new org.fusesource.restygwt.client.ResponseFormatException(
									"Response was NOT a valid JSON document", __e);
						}
					}
				});
	}
}
```

In maven I'm using the following dependencies:
```xml
        <dependency>
			<groupId>org.fusesource.restygwt</groupId>
			<artifactId>restygwt</artifactId>
			<version>2.0.3</version>
		</dependency>
		<dependency>
			<groupId>org.glassfish.jersey.bundles</groupId>
			<artifactId>jaxrs-ri</artifactId>
			<version>2.17</version>
		</dependency>
```

I've tried to change `List<User>` to `User`, then it compiles well.

Could you please help me to find an issue or to find a workaround?

Thanks!

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2009-2012 the original author or authors.
+ * Copyright (C) 2009-2015 the original author or authors.
  * See the notice.md file distributed with this work for additional
  * information regarding copyright ownership.
  *
@@ -298,7 +298,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
         	if(iface != null && REST_SERVICE_TYPE.isAssignableFrom(iface))
         		writeSubresourceLocatorImpl(method);
         	else
-            	writeMethodImpl(method);
+                writeMethodImpl(method, options);
         }
     }
 
@@ -380,7 +380,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                " : com.google.gwt.http.client.URL.encodePathSegment(" + expr + ")))+\"");
     }
 
-    private void writeMethodImpl(JMethod method) throws UnableToCompleteException {
+    private void writeMethodImpl(JMethod method, Options classOptions) throws UnableToCompleteException {
         boolean returnRequest = false;
         if (method.getReturnType() != JPrimitiveType.VOID) {
             if (!method.getReturnType().getQualifiedSourceName().equals(Request.class.getName()) &&
@@ -394,7 +394,6 @@ public class RestServiceClassCreator extends BaseSourceCreator {
         Json jsonAnnotation = getAnnotation(source, Json.class);
         final Style classStyle = jsonAnnotation != null ? jsonAnnotation.style() : Style.DEFAULT;
 
-        Options classOptions = getAnnotation(source, Options.class);
         Options options = getAnnotation(method, Options.class);
 
         p(method.getReadableDeclaration(false, false, false, false, true) + " {").i(1);

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -490,7 +490,12 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                 acceptTypeBuiltIn = "CONTENT_TYPE_XML";
             }
 
-            p("final " + METHOD_CLASS + " __method =");
+            // Handle JSONP specific configuration...
+            JSONP jsonpAnnotation = getAnnotation(method, JSONP.class);
+
+            final boolean isJsonp = restMethod.equals(METHOD_JSONP) && jsonpAnnotation != null;
+
+            p("final " + (isJsonp ? JSONP_METHOD_CLASS : METHOD_CLASS) + " __method =");
 
             p("getResource()");
             if (pathExpression != null) {
@@ -510,20 +515,16 @@ public class RestServiceClassCreator extends BaseSourceCreator {
             // example: .get()
             p("." + restMethod + "();");
 
-            // Handle JSONP specific configuration...
-            JSONP jsonpAnnotation = getAnnotation(method, JSONP.class);
-
-            final boolean isJsonp = restMethod.equals(METHOD_JSONP) && jsonpAnnotation!=null;
             if( isJsonp ) {
                 if (returnRequest && !method.getReturnType().getQualifiedSourceName().equals(JsonpRequest.class.getName())) {
                     getLogger().log(ERROR, "Invalid rest method. JSONP method must have void or JsonpRequest return types: " + method.getReadableDeclaration());
                     throw new UnableToCompleteException();
                 }
                 if( jsonpAnnotation.callbackParam().length() > 0 ) {
-                    p("(("+JSONP_METHOD_CLASS+")__method).callbackParam("+wrap(jsonpAnnotation.callbackParam())+");");
+                    p("__method.callbackParam("+wrap(jsonpAnnotation.callbackParam())+");");
                 }
                 if( jsonpAnnotation.failureCallbackParam().length() > 0 ) {
-                    p("(("+JSONP_METHOD_CLASS+")__method).failureCallbackParam("+wrap(jsonpAnnotation.failureCallbackParam())+");");
+                    p("__method.failureCallbackParam("+wrap(jsonpAnnotation.failureCallbackParam())+");");
                 }
             } else {
                 if (returnRequest && !method.getReturnType().getQualifiedSourceName().equals(Request.class.getName())) {
@@ -675,7 +676,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                 // TODO: shouldn't we also have a cach in here?
                 p(returnRequest(returnRequest,isJsonp) + "__method.send(" + callbackArg.getName() + ");");
             } else if ( isJsonp ){
-                    p(returnRequest(returnRequest,isJsonp) + "((" + JSONP_METHOD_CLASS + ")__method).send(new " + ABSTRACT_ASYNC_CALLBACK_CLASS + "<" + resultType.getParameterizedQualifiedSourceName() + ">((" + JSONP_METHOD_CLASS + ")__method, "
+                    p(returnRequest(returnRequest,isJsonp) + "__method.send(new " + ABSTRACT_ASYNC_CALLBACK_CLASS + "<" + resultType.getParameterizedQualifiedSourceName() + ">(__method, "
                                     + callbackArg.getName() + ") {").i(1);
                     {
                         p("protected " + resultType.getParameterizedQualifiedSourceName() + " parseResult(" + JSON_VALUE_CLASS + " result) throws Exception {").i(1);

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -688,7 +688,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                                 p("try {").i(1);
                                 {
                                     if(resultType.isAssignableTo(locator.LIST_TYPE)){
-                                        p("result = new " + JSON_ARRAY_CLASS + "(result.getJavaScriptObject());");
+                                        p("result = new " + JSON_ARRAY_CLASS + "(((" + JSON_OBJECT_CLASS + ")result).getJavaScriptObject());");
                                     }
                                     jsonAnnotation = getAnnotation(method, Json.class);
                                     Style style = jsonAnnotation != null ? jsonAnnotation.style() : classStyle;

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/util/AnnotationUtils.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/util/AnnotationUtils.java
@@ -20,7 +20,7 @@ public class AnnotationUtils {
             T ann = classType.getAnnotation(annotationType);
             if (ann == null) {
                 for (Annotation metaAnn : classType.getAnnotations()) {
-                    //only return a custom annotation if it contains the JacksonAnnotationsInside annotation
+                    // only return a custom annotation if it contains the JacksonAnnotationsInside annotation
                     if (metaAnn.annotationType().getAnnotation(JacksonAnnotationsInside.class) != null) {
                         ann = metaAnn.annotationType().getAnnotation(annotationType);
                         if (ann != null) {
@@ -45,7 +45,7 @@ public class AnnotationUtils {
         } else {
             List<JClassType> intefaces = Arrays.asList(classType.getImplementedInterfaces());
 
-            for (JClassType itf: intefaces) {
+            for (JClassType itf : intefaces) {
                 annotation = getClassAnnotation(itf, annotationType);
                 if (annotation != null) {
                     return annotation;
@@ -61,7 +61,7 @@ public class AnnotationUtils {
      * <br>
      * Works like {@link JClassType#findAnnotationInTypeHierarchy(Class)} but returns all annotations in the type hierarchy.
      * 
-     * @author Ralf Sommer <ralf.sommer.dev@gmail.com>
+     * @author Ralf Sommer {@literal <ralf.sommer.dev@gmail.com>}
      * 
      * @param classType
      * @return annotations

--- a/restygwt/src/test/java/org/fusesource/restygwt/GwtCompleteTestSuite.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/GwtCompleteTestSuite.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2009-2012 the original author or authors.
+ * Copyright (C) 2009-2015 the original author or authors.
  * See the notice.md file distributed with this work for additional
  * information regarding copyright ownership.
  *
@@ -18,7 +18,6 @@
 
 package org.fusesource.restygwt;
 
-
 import junit.framework.Test;
 import junit.framework.TestCase;
 
@@ -32,6 +31,7 @@ import org.fusesource.restygwt.client.basic.FlakyTestGwt;
 import org.fusesource.restygwt.client.basic.FormParamTestGwt;
 import org.fusesource.restygwt.client.basic.GenericsTestGwt;
 import org.fusesource.restygwt.client.basic.JsonCreatorWithSubtypes;
+import org.fusesource.restygwt.client.basic.JsonpTestGwt;
 import org.fusesource.restygwt.client.basic.ParameterizedTypeDTO;
 import org.fusesource.restygwt.client.basic.ParameterizedTypeServiceInterfaces;
 import org.fusesource.restygwt.client.basic.PathParamTestGwt;
@@ -47,7 +47,6 @@ import org.fusesource.restygwt.client.codec.MapInRestServiceEncoderDecoderTestGw
 import org.fusesource.restygwt.client.complex.JsonTypeIdResolver;
 
 import com.google.gwt.junit.tools.GWTTestSuite;
-
 
 /**
  *
@@ -68,12 +67,11 @@ import com.google.gwt.junit.tools.GWTTestSuite;
  */
 public class GwtCompleteTestSuite extends TestCase {
 
-
     /**
      * @return the suite of that module
      */
     public static Test suite() {
-        GWTTestSuite suite = new GWTTestSuite("all GwtTestCases but AnnotationResolver" );
+        GWTTestSuite suite = new GWTTestSuite("all GwtTestCases but AnnotationResolver");
 
         suite.addTestSuite(BasicTestGwt.class);
         // keep the cache-callback at the beginning to get it pass
@@ -82,13 +80,13 @@ public class GwtCompleteTestSuite extends TestCase {
         suite.addTestSuite(InnerClassesEncoderDecoderTestGwt.class);
         suite.addTestSuite(MapInRestServiceEncoderDecoderTestGwt.class);
         suite.addTestSuite(EncoderDecoderTestGwt.class);
-        
+
         suite.addTestSuite(FlakyTestGwt.class);
         suite.addTestSuite(TimeoutTestGwt.class);
         suite.addTestSuite(CachingTestGwt.class);
         suite.addTestSuite(ResourceTestGwt.class);
         suite.addTestSuite(ResourcePassesHeadersTestGwt.class);
-        suite.addTestSuite(VolatileQueueableCacheStorageTestGwt.class);       
+        suite.addTestSuite(VolatileQueueableCacheStorageTestGwt.class);
         suite.addTestSuite(FailingTestGwt.class);
         suite.addTestSuite(GenericsTestGwt.class);
         suite.addTestSuite(ParameterizedTypeDTO.class);
@@ -101,10 +99,9 @@ public class GwtCompleteTestSuite extends TestCase {
         suite.addTestSuite(FormParamTestGwt.class);
         suite.addTestSuite(DirectRestServiceTestGwt.class);
         suite.addTestSuite(ConfiguredServiceTestGwt.class);
-        
-        //TODO not sure what a jsonp servlet looks like, i.e. the one used by the test just
-        // produces error on the javascript callback
-        //suite.addTestSuite(JsonpTestGwt.class);
+
+        suite.addTestSuite(JsonpTestGwt.class);
+
         return suite;
     }
 }

--- a/restygwt/src/test/java/org/fusesource/restygwt/JsonpTestGwt.gwt.xml
+++ b/restygwt/src/test/java/org/fusesource/restygwt/JsonpTestGwt.gwt.xml
@@ -1,0 +1,35 @@
+<!--
+
+    Copyright (C) 2009-2015 the original author or authors.
+    See the notice.md file distributed with this work for additional
+    information regarding copyright ownership.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+
+<module>
+    <inherits name='com.google.gwt.user.User' />
+    <inherits name='com.google.gwt.logging.Logging'/>
+    <inherits name='org.fusesource.restygwt.RestyGWT'/>
+
+    <!-- Logging Configuration -->
+    <set-property name="gwt.logging.enabled" value="TRUE"/>
+    <set-property name="gwt.logging.logLevel" value="ALL"/>
+
+    <servlet path='/jsonp/*' class='org.fusesource.restygwt.server.basic.JsonpTestGwtServlet' />
+    <servlet path='/ui/*' class='org.fusesource.restygwt.server.basic.JsonpTestGwtServlet' />
+
+    <source path='client'/>
+    <source path='example/client'/>
+</module>

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/basic/JsonpService.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/basic/JsonpService.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2009-2012 the original author or authors.
+ * Copyright (C) 2009-2015 the original author or authors.
  * See the notice.md file distributed with this work for additional
  * information regarding copyright ownership.
  *
@@ -18,12 +18,15 @@
 
 package org.fusesource.restygwt.client.basic;
 
-import com.google.gwt.jsonp.client.JsonpRequest;
 import java.util.List;
+
+import javax.ws.rs.Path;
 
 import org.fusesource.restygwt.client.JSONP;
 import org.fusesource.restygwt.client.MethodCallback;
 import org.fusesource.restygwt.client.RestService;
+
+import com.google.gwt.jsonp.client.JsonpRequest;
 
 /**
  *
@@ -32,20 +35,19 @@ import org.fusesource.restygwt.client.RestService;
  */
 public interface JsonpService extends RestService {
 
-
-    @JSONP(callbackParam = "foo")
+    @JSONP
     public void someJsonp(MethodCallback<ExampleDto> callback);
-    
+
     /**
-     * Used to make sure the generator handles JsonpRequest result. 
+     * Used to make sure the generator handles JsonpRequest result.
      * It can be used to cancel requests.
      */
-    @JSONP(callbackParam = "foo")
-    public JsonpRequest<?> someCancelableJsonp(MethodCallback<ExampleDto> callback);
+    @JSONP
+    public JsonpRequest<ExampleDto> someCancelableJsonp(MethodCallback<ExampleDto> callback);
 
-    // TODO that list generates broken code which does not compile
-//    @JSONP(callbackParam = "list") // param name used by test servlet produce list
-//    public void someJsonpWithList(MethodCallback<List<ExampleDto>> callback);
+    @Path("list")
+    @JSONP
+    public void someJsonpWithList(MethodCallback<List<ExampleDto>> callback);
 
     @JSONP(callbackParam = "null") // param name used by test servlet produce null
     public void someOtherJsonp(MethodCallback<ExampleDto> callback);

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/basic/JsonpTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/basic/JsonpTestGwt.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2009-2012 the original author or authors.
+ * Copyright (C) 2009-2015 the original author or authors.
  * See the notice.md file distributed with this work for additional
  * information regarding copyright ownership.
  *
@@ -18,74 +18,74 @@
 
 package org.fusesource.restygwt.client.basic;
 
+import java.util.List;
+
 import org.fusesource.restygwt.client.Method;
 import org.fusesource.restygwt.client.MethodCallback;
 import org.fusesource.restygwt.client.Resource;
 import org.fusesource.restygwt.client.RestServiceProxy;
-import org.junit.Test;
 
 import com.google.gwt.core.client.GWT;
 import com.google.gwt.jsonp.client.JsonpRequest;
+import com.google.gwt.junit.DoNotRunWith;
+import com.google.gwt.junit.Platform;
 import com.google.gwt.junit.client.GWTTestCase;
+import com.google.gwt.user.client.Timer;
 
 /**
  *
- *
  * @author <a href="mailto:mail@raphaelbauer.com">rEyez</<a>
+ * @author Ralf Sommer {@literal <ralf.sommer.dev@gmail.com>}
  *
  */
 public class JsonpTestGwt extends GWTTestCase {
 
-    private static final String URI_PATH = "ui/jsonp-service";
-
     @Override
     public String getModuleName() {
-        return "org.fusesource.restygwt.BasicTestGwt";
+        return "org.fusesource.restygwt.JsonpTestGwt";
+    }
+
+    JsonpService service;
+
+    @Override
+    protected void gwtSetUp() throws Exception {
+        // configure RESTY
+        Resource resource = new Resource(GWT.getModuleBaseURL() + "jsonp");
+
+        service = GWT.create(JsonpService.class);
+        ((RestServiceProxy) service).setResource(resource);
     }
 
     public void testDefaultFunction() {
-
-        //configure RESTY
-        Resource resource = new Resource(GWT.getModuleBaseURL() + URI_PATH);
-
-
-        JsonpService service = GWT.create(JsonpService.class);
-        ((RestServiceProxy) service).setResource(resource);
+        // wait... we are in async testing...
+        delayTestFinish(5000);
 
         service.someJsonp(new MethodCallback<ExampleDto>() {
-
             @Override
             public void onSuccess(Method method, ExampleDto response) {
-
                 assertEquals(response.name, "myName");
                 finishTest();
-
             }
 
             @Override
             public void onFailure(Method method, Throwable exception) {
                 fail();
-
             }
         });
+    }
 
-        // wait... we are in async testing...
+    /**
+     * HtmlUnit execute synchronously. Should be async.
+     * <p>
+     * Run manually with -Dgwt.args="-prod -runStyle Manual:1" works as expected
+     * 
+     * @see <a href="https://github.com/gwtproject/gwt/blob/138c60c7625a9403f34bd9616cea483fbdbeb2f0/user/test/com/google/gwt/jsonp/client/JsonpRequestTest.java#L191">GWT JsonpRequestTest.java</a>
+     */
+    @DoNotRunWith(Platform.HtmlUnitBug)
+    public void testCancel() {
         delayTestFinish(10000);
 
-    }
-    
-    @Test
-    public void testCancel() {
-
-        //configure RESTY
-        Resource resource = new Resource(GWT.getModuleBaseURL() + URI_PATH);
-
-
-        JsonpService service = GWT.create(JsonpService.class);
-        ((RestServiceProxy) service).setResource(resource);
-
-        JsonpRequest<?> request = service.someCancelableJsonp(new MethodCallback<ExampleDto>() {
-
+        JsonpRequest<ExampleDto> request = service.someCancelableJsonp(new MethodCallback<ExampleDto>() {
             @Override
             public void onSuccess(Method method, ExampleDto response) {
                 fail();
@@ -94,45 +94,54 @@ public class JsonpTestGwt extends GWTTestCase {
             @Override
             public void onFailure(Method method, Throwable exception) {
                 fail();
-
             }
         });
-        
+
         request.cancel();
 
-        // wait... we are in async testing...
-        delayTestFinish(10000);
+        // after waiting for 5 seconds assume, that the request has been canceled successfully
+        new Timer() {
+            @Override
+            public void run() {
+                finishTest();
+            }
+        }.schedule(5000);
     }
 
-//    public void testListFunction() {
-//
-//        //configure RESTY
-//        Resource resource = new Resource(GWT.getModuleBaseURL() + URI_PATH);
-//
-//
-//        JsonpService service = GWT.create(JsonpService.class);
-//        ((RestServiceProxy) service).setResource(resource);
-//
-//        service.someJsonpWithList(new MethodCallback<List<ExampleDto>>() {
-//
-//            @Override
-//            public void onSuccess(Method method, List<ExampleDto> response) {
-//
-//                assertEquals(response.get(0).name, "myName");
-//                finishTest();
-//
-//            }
-//
-//            @Override
-//            public void onFailure(Method method, Throwable exception) {
-//                fail();
-//
-//            }
-//        });
-//
-//        // wait... we are in async testing...
-//        delayTestFinish(10000);
-//
-//    }
+    public void testNullResult() {
+        delayTestFinish(5000);
+
+        service.someOtherJsonp(new MethodCallback<ExampleDto>() {
+            @Override
+            public void onSuccess(Method method, ExampleDto response) {
+                assertNull(response);
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail();
+            }
+        });
+    }
+
+    public void testListFunction() {
+        // wait... we are in async testing...
+        delayTestFinish(5000);
+
+        service.someJsonpWithList(new MethodCallback<List<ExampleDto>>() {
+            @Override
+            public void onSuccess(Method method, List<ExampleDto> response) {
+                assertEquals(response.get(0).name, "myName");
+                assertEquals(response.get(1).name, "myName2");
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail();
+            }
+        });
+    }
 
 }

--- a/restygwt/src/test/java/org/fusesource/restygwt/server/basic/JsonpTestGwtServlet.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/server/basic/JsonpTestGwtServlet.java
@@ -1,0 +1,57 @@
+/**
+ * Copyright (C) 2009-2015 the original author or authors.
+ * See the notice.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fusesource.restygwt.server.basic;
+
+import java.io.IOException;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Simple servlet that answer jsonp test requests.
+ *
+ * @author Ralf Sommer {@literal <ralf.sommer.dev@gmail.com>}
+ *
+ */
+public class JsonpTestGwtServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 4308269541594795760L;
+
+    private static final String DUMMY_RESPONSE = "{\"name\":\"myName\"}";
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/javascript");
+        String responseFunction = request.getParameter("callback");
+        if (null != responseFunction) {
+            if (request.getRequestURI().endsWith("/list")) {
+                response.getWriter().print(responseFunction + "([" + DUMMY_RESPONSE + ",{\"name\":\"myName2\"}" + "]);");
+            } else {
+                response.getWriter().print(responseFunction + "(" + DUMMY_RESPONSE + ");");
+            }
+        } else {
+            responseFunction = request.getParameter("null");
+            if (null != responseFunction) {
+                response.getWriter().print(responseFunction + "();");
+            }
+        }
+    }
+
+}

